Marlin/src/feature/pause.cpp

```diff
@@ -298,48 +331,104 @@ void resume_print(const float &load_length/*=0*/, const float &initial_extrude_l
+    if (swap_spool_disabled)lcd_advanced_pause_show_message(ADVANCED_PAUSE_MESSAGE_RESUME);
   #endif
-
-  // Set extruder to saved position
-  planner.set_e_position_mm((current_position[E_AXIS] = resume_position[E_AXIS]));
-
+	
+  // Stop blowing after delay
+  #ifdef ADVANCED_PAUSE_DWELL
+    delay(ADVANCED_PAUSE_DWELL);
+  #endif
+  #if (ADVANCED_PAUSE_FANSPEED >0 && ADVANCED_PAUSE_FAN < FAN_COUNT)
+    fanSpeeds[ADVANCED_PAUSE_FAN]=0 ;
+  #endif  
+ 
+  // Intelligent Retractation before return to print
+  // If resume position negative then travel back Retraction = resume_position
+  // Rare but possible : can go on pause with retracted/swapped status or negative position
+  #ifdef PAUSE_PARK_RETRACT_LENGTH
+    #if ENABLED(FWRETRACT)
+	  //If retracted/swapped goto negative position
+	  if (fwretract.retracted_swap[active_extruder]) {
+		const float tmp_length=-fwretract.swap_retract_length;
+	    const float tmp_feed=fwretract.retract_feedrate_mm_s;
+	    do_pause_e_move(tmp_length, tmp_feed); 
+	  }
+	  // If retracted or not : normal fwretract used
+	  else {
+	   const float tmp_length=-fwretract.retract_length;  
+       const float tmp_feed=fwretract.retract_feedrate_mm_s;   
+       do_pause_e_move(tmp_length, tmp_feed); 
+	 }
+    #else
+	  //If negative and more than PAUSE_PARK_RETRACT_LENGTH goto resume_position 
+	  if (resume_position[E_AXIS]<(-PAUSE_PARK_RETRACT_LENGTH)) do_pause_e_move(resume_position[E_AXIS], PAUSE_PARK_RETRACT_FEEDRATE);
+      else do_pause_e_move(-PAUSE_PARK_RETRACT_LENGTH, PAUSE_PARK_RETRACT_FEEDRATE);
+    #endif
+  #endif	
+ 
   // Move XY to starting position, then Z
   do_blocking_move_to_xy(resume_position[X_AXIS], resume_position[Y_AXIS], PAUSE_PARK_XY_FEEDRATE);
-  do_blocking_move_to_z(resume_position[Z_AXIS], PAUSE_PARK_Z_FEEDRATE);
 
+   //Set Z_AXIS to saved position  
+  do_blocking_move_to_z(resume_position[Z_AXIS], PAUSE_PARK_Z_FEEDRATE);
+  
+  // Intelligent Retractation Recovery
+  // Keep applied retractation as resume position if retracted before goto pause
+  #ifdef PAUSE_PARK_RETRACT_LENGTH
+    #if ENABLED(FWRETRACT)
+	  //if not retracted/swapped before , then recover to zero
+	  if (!fwretract.retracted_swap[active_extruder] && !fwretract.retracted[active_extruder]){
+		const float tmp_length=fwretract.swap_retract_recover_length;
+	    const float tmp_feed=  fwretract.swap_retract_recover_feedrate_mm_s;
+	    do_pause_e_move(tmp_length, tmp_feed); 
+	  }	  			
+	#else    
+	  // If resume_position negative goto resume_position
+      if (resume_position[E_AXIS]<0) do_pause_e_move(resume_position[E_AXIS], PAUSE_PARK_RETRACT_FEEDRATE);
+	  // if positive soft recovery to zero for sure print restart
+      else do_pause_e_move(PAUSE_PARK_RETRACT_LENGTH,15); 
+    #endif
+  #endif
+  
+  //Now all extrusion positions are resumed and ready to be confirmed
+  //Set extruder to saved position
+  planner.set_e_position_mm((current_position[E_AXIS] = resume_position[E_AXIS]));
+  
```

Comment:
Now we are on the print surface , and we resume with intelligence
and once resumed at the exact old position , we can apply the position on planner